### ISSUE TITLE
Kibana plugin installations fail due to not finding package's Node modules

### ISSUE THREAD

**Issue by ches** [Homebrew/homebrew#48838]
```
$ kibana plugin --install elastic/sense
Installing sense
Attempting to transfer from https://download.elastic.co/elastic/sense/sense-latest.tar.gz
Transferring 318849 bytes....................
Transfer complete
Extracting plugin archive
Extraction complete
Optimizing and caching browser bundles...
Plugin installation was unsuccessful due to error "Cannot find module 'joi'"
```

For a survey of how people don't browse issues before reporting :grin:, see:

  - elastic/sense#93
  - elastic/sense#94
  - elastic/sense#96

A quick and dirty kludge to fix this is to symlink the `node_modules` alongside the plugin directory created and symlinked by the formula:

```patch
diff --git a/Library/Formula/kibana.rb b/Library/Formula/kibana.rb
index 268d251..5e9f4b6 100644
--- a/Library/Formula/kibana.rb
+++ b/Library/Formula/kibana.rb
@@ -65,6 +65,7 @@ class Kibana < Formula

     (var/"lib/kibana/installedPlugins").mkpath
     ln_s var/"lib/kibana/installedPlugins", prefix/"installedPlugins"
+    ln_s prefix/"node_modules", var/"lib/kibana/node_modules"
   end

   plist_options :manual => "kibana"
```

But that probably isn't the *right* fix. I haven't grokked how the plugin installation works yet (beyond a hand-wavy "something npm"), just wanted to get the issue reported.

It would be nice to leave plugins in place when users upgrade Kibana, which must be the intent of the `installedPlugins` symlink that's used now. If the plugins share `node_modules` with Kibana itself, though, and have implicit dependencies on libs Kibana uses—the `sense` plugin does not explicitly declare an npm dep on `joi` or anything else, which might be a discussion to raise upstream—then this might become a mess across upgrades.

#46593 is relevant for history of the package-local copy of Node.

**Comment by mikemcquaid** [comment]
> But that probably isn't the right fix. I haven't grokked how the plugin installation works yet (beyond a hand-wavy "something npm"), just wanted to get the issue reported.

FWIW that seems fine to me. CC @jasontedor who may have better ideas.

**Comment by jasontedor** [comment]
Thanks for reporting @ches and thanks for the ping @mikemcquaid.

> It would be nice to leave plugins in place when users upgrade Kibana, which must be the intent of the installedPlugins symlink that's used now.

Yes, that was the intent.

> If the plugins share node_modules with Kibana itself, though, and have implicit dependencies on libs Kibana uses—the sense plugin does not explicitly declare an npm dep on joi or anything else, which might be a discussion to raise upstream—then this might become a mess across upgrades.

Yes, exactly. Since it is just going to create a mess, I think that the best approach is to just keep the `installedPlugins` folder inside the Kibana keg. This does mean that plugins will be lost across upgrades, but those plugins will require upgrades anyway (and the upgrade process is just remove followed by install). This is the same as the Elasticsearch formula where since Elasticsearch 2.0.0, plugins are tightly bound to the version of Elasticsearch so they also can not be preserved upon upgrade.

There is work underway right now to change the way that plugins are handled across the Elastic stack so this will probably need to be revisited when that work is completed.

I opened #48951. I included a caveat about the plugins situation.
